doc/operations/configuration.md
# Configuration

Configuration of the environment is largely performed via [`env.sh`](../../env.sh) in the root
directory.

In order to decouple components, an effort is made to use native Kubernetes
components and only provide configuration as to the opinion of each item and
those chosen by the operator. 

## Configuration Items

`env.sh` is sourced by other code in the project, and is the central point of
configuration. By default, it assumes `minikube` is used and as such searches
the user PATH for `minikube` and other utilities required for the runtime such
as `helm`, `git`, `virtualenv`, `yq`, and `jq`. It then aliases `kubectl` as
`minikube kubectl`. Users may edit this piece of code for any custom uses of
`kubectl`.

`env.sh` supports [OpenShift Local](https://developers.redhat.com/products/openshift-local/overview) via [`crc`](https://crc.dev/crc/). In order to use this capability,
one needs to create a [Red Hat developer account](https://developers.redhat.com/blog/2016/03/31/no-cost-rhel-developer-subscription-now-available#), [download and install](https://console.redhat.com/openshift/create/local), and
[configure the cluster](https://access.redhat.com/documentation/en-us/red_hat_openshift_local/2.5/html/getting_started_guide/configuring_gsg). In order to
use OpenShift Local in the project, `RUNTIME` in `env.sh` needs to be "crc".

`LOGLEVEL` can be set to "INFO" or "DEBUG". This setting is straightforward,
and "DEBUG" increases verbosity of output in the code.

`POD_NET_CIDR` defaults to "172.16.0.0/16" but is configurable by editing this
variable.

`PROJECT_NAMESPACE` is set to "boutique" by default. It is used to manage the
namespace across the project.

In an effort to provide support for TCP proxy for localhost forwarding to
Minikube, `DOCKER_REG_PORT` is set to 5000. For platforms this unnecessary,
the user can ignore or provide custom code to meet their use case.

There are use cases custom images will need to be pushed to a registry, and the
examples included use Minishift registry addon. This is defined as the well
known `DOCKER_HOST` variable. Support for integrated registries such as
OpenShift is provided. The user can modify according to their use case --
`eval $(crc podman-env)`. In this case, `IGNORE_DOCKER_CONFIG` needs to be set
as `true`.


env.sh
#!/usr/bin/env bash
#
# Copyright 2022 Justin Cook
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to
# deal in the Software without restriction, including without limitation the
# rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
# sell copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.
#
# These are entries to configure and modify the environment for code in this
# project.
#
# Author: Justin Cook

set -o nounset errexit

# Set the runtime. Supported options are minikube and crc.
export RUNTIME="minikube"
#export RUNTIME="crc"

# Set loglevel to screen. The valid options are INFO and DEBUG
export LOGLEVEL="DEBUG"

# Pod network CIDR
export POD_NET_CIDR="172.16.0.0/16"

# Which namespace will the project reside?
export PROJECT_NAMESPACE="boutique"

# Is the environment running?
export RUNNING=false

_exit_() {
  local lc="$BASH_COMMAND" rc=$?
  if [ "${LOGLEVEL}" = "DEBUG" ]
  then
    echo "Exited with code [$rc]: [$lc]"
  fi
  tty -s
}

trap _exit_ EXIT

if [ "${LOGLEVEL-}" = "DEBUG" ]
then
  set -x
  REDIRECT=""
else
  REDIRECT=">/dev/null"
fi

# Check if all the necessary utilities are available or exit
# `kubectl` is unnecessary in this context as it is later aliased
check_dependencies() {
  for cmd in "${RUNTIME}" "helm" "git" "virtualenv" "yq" "jq"
  do
    if ! command -v "${cmd}" &> /dev/null
    then
        echo "command: ${cmd} could not be found"
        exit 1
    fi
  done
}

# Changes in versioning with an independently installed kubectl utility can
# cause unexpected outcomes with API access. The assumption is Kubernetes is on
# Minikube or OpenShift Local.
check_platform() {
  if [ "${RUNTIME}" == "minikube" ]
  then
    if (which minikube && minikube status)
    then
      RUNNING=true
    fi
    alias kubectl="minikube kubectl --"
  elif [ "${RUNTIME}" = "crc" ]
  then
    if (which crc && crc status)
    then
      RUNNING=true
      #shellcheck disable=SC2046
      eval $(crc oc-env)
    fi
    alias kubectl="oc"
  else
    alias kubectl="kubectl --kubeconfig=kubeconfig --insecure-skip-tls-verify=true"
    RUNNING=true
  fi
}

# Create the namespace
create_namespace(){
  if ${RUNNING} && [ "${RUNTIME}" = "crc" ]
  then
    if ! oc get project "${PROJECT_NAMESPACE}"
    then
      oc new-project "${PROJECT_NAMESPACE}" \
        --description="This is the Google Boutique microservices demo" \
        --display-name="Online Boutique"
    fi
    oc adm policy add-scc-to-user privileged system:serviceaccount:"${PROJECT_NAMESPACE}":default
  elif ${RUNNING}
  then
    kubectl create ns "${PROJECT_NAMESPACE}" --dry-run=client -o yaml | \
    kubectl apply -f -
  fi
}

# Set the default namespace to PROJECT_NAMESPACE
set_default_namespace() {
  if ${RUNNING}
  then
    kubectl config set-context --current --namespace="${PROJECT_NAMESPACE}"
  fi
}

# Prometheus namespace and service names
set_prometheus_names() {
  if [ "${RUNTIME}" = "crc" ]
  then # OpenShift
    export PROMETHEUS_NS="openshift-monitoring"
    export PROMETHEUS_SVC="prometheus-operated"
  else # Rancher
    export PROMETHEUS_NS="cattle-monitoring-system"
    export PROMETHEUS_SVC="rancher-monitoring-prometheus"
  fi
}

# Where is Docker available, and what port should be forwarded?
set_docker_env() {
  if ${RUNNING} && [ "${RUNTIME}" = "crc" ]
  then
    #shellcheck disable=SC2046
    eval $(crc podman-env)
    export IGNORE_DOCKER_CONFIG=true
  else
    export DOCKER_HOST="tcp://localhost:2375"
    export DOCKER_REG_PORT="5000"
    export IGNORE_DOCKER_CONFIG=false
  fi
}

CMDS=$(cat - <<__EOF__
check_dependencies
check_platform
create_namespace
set_default_namespace
set_prometheus_names
set_docker_env
__EOF__
)

for cmd in ${CMDS}
do
  #shellcheck disable=SC2046
  eval $(printf '%s' "${cmd} ${REDIRECT-}")
done